forge/src/main/java/me/lucko/luckperms/forge/ForgeSenderFactory.java

```diff
@@ -0,0 +1,110 @@
+import me.lucko.luckperms.common.query.QueryOptionsImpl;
+import me.lucko.luckperms.common.sender.Sender;
+import me.lucko.luckperms.common.sender.SenderFactory;
+import me.lucko.luckperms.common.verbose.VerboseCheckTarget;
+import me.lucko.luckperms.common.verbose.event.CheckOrigin;
+import me.lucko.luckperms.forge.bridge.server.level.ServerPlayerBridge;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer;
+import net.luckperms.api.util.Tristate;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.server.level.ServerPlayer;
+
+import java.util.Locale;
+import java.util.UUID;
+
+public class ForgeSenderFactory extends SenderFactory<LPForgePlugin, CommandSourceStack> {
+    public ForgeSenderFactory(LPForgePlugin plugin) {
+        super(plugin);
+    }
+
+    @Override
+    protected UUID getUniqueId(CommandSourceStack commandSource) {
+        if (commandSource.getEntity() != null) {
+            return commandSource.getEntity().getUUID();
+        }
+        return Sender.CONSOLE_UUID;
+    }
+
+    @Override
+    protected String getName(CommandSourceStack commandSource) {
+        String name = commandSource.getTextName();
+        if (commandSource.getEntity() != null && name.equals("Server")) {
+            return Sender.CONSOLE_NAME;
+        }
+        return name;
+    }
+
+    @Override
+    protected void sendMessage(CommandSourceStack sender, Component message) {
+        Locale locale;
+        if (sender.getEntity() instanceof ServerPlayer) {
+            locale = ((ServerPlayerBridge) sender.getEntity()).bridge$getLocale();
+        } else {
+            locale = null;
+        }
+
+        sender.sendSuccess(toNativeText(TranslationManager.render(message, locale)), false);
+    }
+
+    @Override
+    protected Tristate getPermissionValue(CommandSourceStack commandSource, String node) {
+        if (commandSource.getEntity() instanceof ServerPlayer) {
+            return ((ServerPlayerBridge) commandSource.getEntity()).bridge$hasPermission(node);
+        }
+
+        VerboseCheckTarget target = VerboseCheckTarget.internal(commandSource.getTextName());
+        getPlugin().getVerboseHandler().offerPermissionCheckEvent(CheckOrigin.PLATFORM_API_HAS_PERMISSION, target, QueryOptionsImpl.DEFAULT_CONTEXTUAL, node, TristateResult.UNDEFINED);
+        getPlugin().getPermissionRegistry().offer(node);
+        return Tristate.UNDEFINED;
+    }
+
+    @Override
+    protected boolean hasPermission(CommandSourceStack commandSource, String node) {
+        return getPermissionValue(commandSource, node).asBoolean();
+    }
+
+    @Override
+    protected void performCommand(CommandSourceStack sender, String command) {
+        sender.getServer().getCommands().performCommand(sender, command);
+    }
+
+    @Override
+    protected boolean isConsole(CommandSourceStack sender) {
+        return sender.getEntity() == null;
```

Comment:
Are you sure this is correct?

`commandSource.getEntity() != null && name.equals("Server")` is used above to check for console? Although I'm not convinced that's the correct way either..